### PR TITLE
[test-framework] Integrating local Pilot and Envoy

### PR DESCRIPTION
**Pull request by nmittler**
- Updated configuration of Envoy to use discovery v2 API on Pilot

- Adding a discovery proxy to allow interception of responses from
Pilot. This will provide the ability to modify the Envoy configs to
support local testing.

- Updated the agent test to incorporate a local pilot. Also starting
to hash out how to determine whether or not an Envoy has received
a configuration update.

### CODE REVIEW

File: pkg/test/local/proxy/envoy/agent/agent.go [hunk 1]
```diff
@@ -119,30 +174,37 @@ func (a *Agent) stopService() error {
...
-		Ports:       a.ports,
-		tmpDir:      a.Config.TmpDir,
+	nodeID := a.generateServiceNode()
+	a.envoyConfig, err = (&envoyConfigBuilder{
+		ServiceName:   a.Config.ServiceName,
+		NodeID:        nodeID,
+		Cluster:       serviceCluster,
+		AdminPort:     a.proxyAdminPort,
+		Ports:         a.ports,
+		DiscoveryHost: "127.0.0.1",
```
Comment (ozevren) [inline pkg/test/local/proxy/envoy/agent/agent.go]: Should this use "ipAddress" const?

File: pkg/test/local/proxy/envoy/discovery/interceptor.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+var (
+	// Identity is a global instance of the identity interceptor.
+	Identity = &IdentityResponseInterceptor{}
+)
+
+// ResponseInterceptor is an interceptor of discovery responses from the server, which allows the application
+// to modify the response returned to the client.
+type ResponseInterceptor interface {
+	InterceptDiscoveryResponse(resp *xdsapi.DiscoveryResponse) (*xdsapi.DiscoveryResponse, error)
```
Comment (ozevren) [inline pkg/test/local/proxy/envoy/discovery/interceptor.go]: Nit: it looks like this simply be a lambda.

File: pkg/test/local/proxy/envoy/agent/agent.go (hunk 1 above)
Comment (nmittler) [reply pkg/test/local/proxy/envoy/agent/agent.go]: Done.

File: pkg/test/local/proxy/envoy/agent/discovery_proxy.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"net"
+
+	"google.golang.org/grpc"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/test/local/proxy/envoy/discovery"
+)
+
+// discoveryProxy acts as a proxy between the local Envoy and Pilot and uses
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/test/local/proxy/envoy/agent/discovery_proxy.go]: This class has been removed.  I don't think this comment applies any longer.

File: pkg/test/local/proxy/envoy/agent/discovery_proxy.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+type discoveryProxy struct {
+	agent      *Agent
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/test/local/proxy/envoy/agent/discovery_proxy.go]: I think the new layout is clearer.  Let's sync up tomorrow and if it's still an issue I'll address it then.

File: pkg/test/local/proxy/envoy/discovery/interceptor.go (hunk 2 above)
Comment (nmittler) [reply pkg/test/local/proxy/envoy/discovery/interceptor.go]: Good point. Done.

File: pkg/test/local/envoy/discovery/filter.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+		return resp, nil
+	}
+)
+
+// FilterFunc is a function that allows an application to filter Envoy discovery responses
+type FilterFunc func(resp *xdsapi.DiscoveryResponse) (*xdsapi.DiscoveryResponse, error)
+
+// Filter proxies requests to an Envoy discovery server (e.g. Pilot) and filters the responses.
+type Filter struct {
+	// DiscoveryAddr is the address of the discovery server
```
Comment (ozevren) [inline pkg/test/local/envoy/discovery/filter.go]: Perhaps make it explicit that this is the address of Filter's discovery service.

Comment (nmittler) [reply pkg/test/local/envoy/discovery/filter.go]: I'm curious what would make this more obvious? The property is already named `DiscoveryAddr`, which seems pretty straight-forward.  Would `DiscoveryServiceAddr` be better in your mind?

Comment (nmittler) [reply pkg/test/local/envoy/discovery/filter.go]: Discussed offline and modified the comment to clarify.

**Pull request merged**